python/pyarrow/tests/test_csv.py

```diff
@@ -1972,6 +1972,33 @@ def test_write_csv_decimal(tmpdir, type_factory):
     assert out.column('col').cast(type) == table.column('col')
 
 
+def test_large_binary_write_to_csv(tmpdir):
+    data_size = int(1E6)
+    file_name = tmpdir / "fixedsize_"+str(data_size)+".csv"
+
+    nparr = np.frombuffer(np.random.randint(65, 91, data_size, 'u1'), 'S4')
+
+    fixed_arr = pa.array(nparr, pa.binary(4))
+    fixed_table = pa.Table.from_arrays([fixed_arr], names=['fixedsize'])
+
+    write_options = WriteOptions(include_header=True, batch_size=2048,
+                                 delimiter='|')
```

Comment:
There is nothing wrong with this, but the custom delimiter seems gratuitous. Is the batch size related to the data size?

Reply:
(if it is, this means you could have a much smaller `data_size` and something like `batch_size = data_size // 4`? also with a comment mentioning why this is important for this test)

Reply:
I'm pretty sure both `batch_size` and `delimiter` were picked up from the code that originally caught the bug: https://github.com/apache/arrow/issues/35901.

The `data_size` seems important for triggering the case (smaller `data_sizes` weren't triggering it). I'll play with `batch_size`, and remove the delimiter.

Reply:
Turns out `batch_size` is not important, either!

Reply:
The `batch_size` doesn't matter much, but for the sake of error reproduction, I used it. But the data size is important. I agree with @anjakefala